mindmeld/components/question_answerer.py

```diff
@@ -451,803 +528,2178 @@ def _generate_mapping_data(embedder_model, embedding_fields):
+                raise TypeError(errmsg)
+
+            if self.data_type == "bool":
+                if not self.is_bool(value):
+                    _raise_error()
+            elif self.data_type == "number":
+                if not self.is_number(value):
+                    _raise_error()
+            elif self.data_type == "string":
+                if self.is_string(value):
+                    value = value.strip()
+                elif self.is_list_of_strings(value):
+                    value = [val.strip() for val in value]
+                else:
+                    _raise_error()
+            elif self.data_type == "date":
+                value = value.strip()
+                if not self.is_date(value):
+                    _raise_error()
+            elif self.data_type == "location":
+                if not self.is_location(value):
+                    _raise_error()
+                # convert it into standard format "37.77,122.41"
+                if isinstance(value, list):  # eg. [37.77, 122.41]
+                    value = ",".join([str(_value) for _value in value])
+                elif isinstance(value, dict):  # eg. {"lat": 37.77, "lon": 122.41}
+                    value = ",".join([str(value["lat"]), str(value["lon"])])
+                elif isinstance(value, str):  # eg. "37.77,122.41"
+                    value = value.strip()
+
+            return value
+
+        def update_resource(self, id2value, has_text_resolver, has_embedding_resolver,
+                            resolver_settings, lazy_clean=False,
+                            app_path=NON_ELASTICSEARCH_INDICES_STORAGE_PATH,
+                            processor_type="keyword"):
+            """
+            Updates a field resource by augmenting/updating latest data and resolvers
+
+            Args:
+                id2value (dict): a mapping between documnet ids & values of the chosen KB field
+                has_text_resolver (bool): If a tfidf resolver is to be created
+                has_embedding_resolver (bool): If a embedder resolver is to be created
+                resolver_settings (dict): a ER- or QA- config with `model_settings` keyword;
+                    used while fitting any resolver
+                lazy_clean (bool, optional): if True, resolvers are fit with clean=True; tagged lazy
+                    because the embedder cache, if cleaned, is cleaned later than the index cache
+                app_path (str, optional): a path to create cache for embedder resolver
+                processor_type (str, optional, "text" or "keyword"): processor for tfidf resolver
+            """
 
-        def validate(self):
+            if not id2value and not self.data_type:
+                return  # else, if required, update resolvers
+
+            # update id2cname and compute hash
+            for _id, value in id2value.items():
+                # ignore null values
+                if not isinstance(value, bool) and not value:
```

Comment:
```suggestion
                if not value and not isinstance(value, bool):
```

Reply:
Consciously put it in reverse order so as to first check if it is a bool, then check for `None`. If we make it other way round and if `value=False` (a meaningful input!), the data is not recorded and ignored. In the way I initially wrote, the input is only ignored if the input is not a boolean and then if it's a null value. Does that make sense?